banyan/src/zstd_dag_cbor_seq.rs

```diff
@@ -248,42 +249,40 @@ impl ZstdDagCborSeq {
     }
 
     /// encrypt using the given key and nonce
-    pub fn encrypt(
-        &self,
-        nonce: &chacha20::XNonce,
-        key: &chacha20::Key,
-    ) -> anyhow::Result<Vec<u8>> {
-        self.clone().into_encrypted(nonce, key)
+    pub fn encrypt(&self, key: &chacha20::Key, mut offset: u64) -> anyhow::Result<Vec<u8>> {
+        self.clone().into_encrypted(key, &mut offset)
     }
 
     /// convert into an encrypted blob, using the given key and nonce
-    pub fn into_encrypted(
-        self,
-        nonce: &chacha20::XNonce,
-        key: &chacha20::Key,
-    ) -> anyhow::Result<Vec<u8>> {
+    pub fn into_encrypted(self, key: &chacha20::Key, offset: &mut u64) -> anyhow::Result<Vec<u8>> {
         let Self { mut data, links } = self;
         // encrypt in place with the key and nonce
-        let mut chacha20 = XChaCha20::new(key, nonce);
+        let mut chacha20 = XChaCha20::new(key, &NONCE.into());
+        let len = data.len();
+        chacha20.seek(*offset);
         chacha20.apply_keystream(&mut data);
         // add the nonce
-        data.extend(nonce.as_slice());
+        data.extend(&offset.to_be_bytes());
+        *offset += len as u64;
```

Comment:
This seems to me like it should be the responsibility of the caller — unless `len` somehow is not known to upper layers (which wasn’t the impression I got).

Reply:
Well, this is just hacked in. I am working on a larger refactoring that will make things clearer. Basically, I really want to avoid having some misunderstanding that leads to offsets being reused.